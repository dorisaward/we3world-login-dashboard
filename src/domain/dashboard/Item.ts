export type Item = {
  /**
   * Unique identifier for the Item
   */
  id: string;

  /**
   * Human readable name of the Item
   */
  name: string;

  /**
   * Human readable description of the Item
   */
  description: string;
};
