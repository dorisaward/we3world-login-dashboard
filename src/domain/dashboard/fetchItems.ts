/**
 * @returns the {@link items} to be displayed
 */
export const fetchItems = async () => {
  try {
    const response = await fetch('/api/items');
    const responseJson = await response.json();
    return responseJson;
  } catch (error) {
    console.error(error);
  }
};
