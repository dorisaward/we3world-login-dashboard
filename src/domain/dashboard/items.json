[
  {
    "id": "0",
    "name": "Item 1",
    "description": "Description aesrdthfjfdhgzxchjvn srdthsrtrdhf srdthfyg"
  },
  {
    "id": "1",
    "name": "Item 2",
    "description": "chgncgnbdhthdt sxchncgbdxgcnb xfrhfyjmfyhm xfgtsrtgsdrg"
  },
  {
    "id": "2",
    "name": "Item 3",
    "description": "zxfhbvcbngnhjfdtj aesrdthfjfdhgzxchjvn srdthsrtrdhf srdthfyg"
  }
]