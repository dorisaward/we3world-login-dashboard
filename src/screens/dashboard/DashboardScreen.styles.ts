import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  container: {
    height: '100%',
    padding: 10,
    flex: 1,
  },
  activityIndicaton: {
    flex: 1,
  },
});
