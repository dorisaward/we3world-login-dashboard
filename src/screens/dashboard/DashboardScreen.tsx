import React, {useEffect, useState} from 'react';
import {ActivityIndicator, Button, FlatList, View} from 'react-native';
import {styles} from './DashboardScreen.styles';
import {Item} from '../../domain/dashboard/Item';
import {DashboardItemView} from './components/DashboardItemView';
import {fetchItems} from '../../domain/dashboard/fetchItems';

type Props = {
  /**
   * Unrenders the dashboard screen and renders the {@link LoginScreen}
   */
  navigateToLoginScreen: () => void;
};

export const DashboardScreen = ({
  navigateToLoginScreen,
}: Props): React.JSX.Element => {
  const [items, setItems] = useState<Item[]>([]);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    fetchItems().then(fetchedItems => {
      setItems(fetchedItems);
      setLoading(false);
    });
  }, []);
  return (
    <View style={styles.container}>
      {loading ? (
        <ActivityIndicator
          testID="dashboard-screen-spinner"
          color={'mediumvioletred'}
          style={styles.activityIndicaton}
          size={'large'}
        />
      ) : (
        <FlatList
          data={items}
          renderItem={DashboardItemView}
          keyExtractor={keyExtractor}
        />
      )}
      <Button title="logout" onPress={navigateToLoginScreen} />
    </View>
  );
};

const keyExtractor = (item: Item) => item.id;
