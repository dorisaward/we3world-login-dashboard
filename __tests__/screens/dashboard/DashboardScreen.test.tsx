import React from 'react';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import {DashboardScreen} from '../../../src/screens/dashboard/DashboardScreen';
import items from '../../../src/domain/dashboard/items.json';
import {fetchItems} from '../../../src/domain/dashboard/fetchItems';

const mockDashboardItemViewText = 'mockDashboardItemViewText';

jest.mock(
  '../../../src/screens/dashboard/components/DashboardItemView',
  () => ({
    DashboardItemView: (props: any) => {
      const RNText = require('react-native').Text;
      return <RNText {...props}>{mockDashboardItemViewText}</RNText>;
    },
  }),
);

jest.mock('../../../src/domain/dashboard/fetchItems', () => ({
  fetchItems: jest.fn().mockResolvedValue(undefined),
}));

it('should render, given items not yet fetched', () => {
  // Given
  const renderable = <DashboardScreen navigateToLoginScreen={jest.fn()} />;

  // When
  const {toJSON, getByTestId} = render(renderable);

  // Then
  expect(getByTestId('dashboard-screen-spinner')).toBeTruthy();
  expect(toJSON()).toMatchSnapshot();
});

it('should render, given items fetched', async () => {
  // Given
  jest.mocked(fetchItems).mockResolvedValueOnce(items);
  const renderable = <DashboardScreen navigateToLoginScreen={jest.fn()} />;
  const {toJSON, queryByTestId, getAllByText} = render(renderable);

  // When
  await waitFor(() =>
    expect(queryByTestId('dashboard-screen-spinner')).toBeFalsy(),
  );

  // Then
  expect(queryByTestId('dashboard-screen-spinner')).toBeFalsy();
  expect(getAllByText(mockDashboardItemViewText)).toHaveLength(3);
  expect(toJSON()).toMatchSnapshot();
});

it('should navigate, given logout button pressed', () => {
  // Given
  const navigateToLoginScreen = jest.fn();
  const renderable = (
    <DashboardScreen navigateToLoginScreen={navigateToLoginScreen} />
  );
  const {getByText} = render(renderable);
  const logoutButton = getByText('logout');

  // When
  fireEvent.press(logoutButton);

  // Then
  expect(navigateToLoginScreen).toHaveBeenCalled();
});
