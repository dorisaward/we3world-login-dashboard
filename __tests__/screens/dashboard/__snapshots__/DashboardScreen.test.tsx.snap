// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render, given items fetched 1`] = `
<View
  style={
    {
      "flex": 1,
      "height": "100%",
      "padding": 10,
    }
  }
>
  <RCTScrollView
    data={
      [
        {
          "description": "Description aesrdthfjfdhgzxchjvn srdthsrtrdhf srdthfyg",
          "id": "0",
          "name": "Item 1",
        },
        {
          "description": "chgncgnbdhthdt sxchncgbdxgcnb xfrhfyjmfyhm xfgtsrtgsdrg",
          "id": "1",
          "name": "Item 2",
        },
        {
          "description": "zxfhbvcbngnhjfdtj aesrdthfjfdhgzxchjvn srdthsrtrdhf srdthfyg",
          "id": "2",
          "name": "Item 3",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text
          index={0}
          item={
            {
              "description": "Description aesrdthfjfdhgzxchjvn srdthsrtrdhf srdthfyg",
              "id": "0",
              "name": "Item 1",
            }
          }
          separators={
            {
              "highlight": [Function],
              "unhighlight": [Function],
              "updateProps": [Function],
            }
          }
        >
          mockDashboardItemViewText
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text
          index={1}
          item={
            {
              "description": "chgncgnbdhthdt sxchncgbdxgcnb xfrhfyjmfyhm xfgtsrtgsdrg",
              "id": "1",
              "name": "Item 2",
            }
          }
          separators={
            {
              "highlight": [Function],
              "unhighlight": [Function],
              "updateProps": [Function],
            }
          }
        >
          mockDashboardItemViewText
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text
          index={2}
          item={
            {
              "description": "zxfhbvcbngnhjfdtj aesrdthfjfdhgzxchjvn srdthsrtrdhf srdthfyg",
              "id": "2",
              "name": "Item 3",
            }
          }
          separators={
            {
              "highlight": [Function],
              "unhighlight": [Function],
              "updateProps": [Function],
            }
          }
        >
          mockDashboardItemViewText
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        logout
      </Text>
    </View>
  </View>
</View>
`;

exports[`should render, given items not yet fetched 1`] = `
<View
  style={
    {
      "flex": 1,
      "height": "100%",
      "padding": 10,
    }
  }
>
  <ActivityIndicator
    color="mediumvioletred"
    size="large"
    style={
      {
        "flex": 1,
      }
    }
    testID="dashboard-screen-spinner"
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        logout
      </Text>
    </View>
  </View>
</View>
`;
